### DISCUSSION TITLE
Should RichLog support colors (and may be other features) on Windows terminal?

### DISCUSSION THREAD

**Discussion by belonesox** [Textualize/textual#3888] [Q&A] [answered]
Since version 1909 Windows support ANSI colors in `cmd` (before — support in `ConEmu` and other advaced windows consoles), and most `textualize` widgets works OK, with colors.

I make `RichLog` working without crash in encodings with 
```
chcp 65001
set PYTHONIOENCODING=UTF-8
```
But does not see any colors (running sample from https://github.com/davidfokkema/playground-textual/blob/main/console_richlog.py )

Is something wrong with my setup, or RichLog should not support Windows at all?

**Comment by TomJGooding** [comment]
Is your question about the `RichLog` generally, or specific to the rather complicated example you've linked?

What do you see if you run this example below?

```python
from rich.text import Text
from textual.app import App, ComposeResult
from textual.widgets import RichLog


class ExampleApp(App):
    def compose(self) -> ComposeResult:
        yield RichLog()

    def on_mount(self) -> None:
        rich_log = self.query_one(RichLog)
        red_text = Text.from_ansi("\033[31mRed Text\033[0m")
        green_bkgd = Text.from_ansi("\033[42mGreen Background\033[0m")
        rich_log.write(red_text)
        rich_log.write(green_bkgd)


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Comment by belonesox** [comment]
Hmm. I see colors

![image](https://github.com/Textualize/textual/assets/1609739/2f4e756c-3eea-40dd-a433-2a7ee3f1dd11)

Even `python -m rich.progress` looks nice (not so nice as in Linux, but with colors and refreshing progress bars).

![image](https://github.com/Textualize/textual/assets/1609739/726c86e1-8dc7-40fe-b50c-37d1110f8f73)



But then I run `console_richlog.py`, I do not see colors in pretty printing and other samples (and definitely something wrong with redrawing progress lines):

![image](https://github.com/Textualize/textual/assets/1609739/b0b8a713-2720-4cb7-85c7-b821f69dacc8)

**Reply by davep** [reply]
Do note that that sample code has nothing to do with the Textual project. In the first instance, if you're having trouble with someone else's code, the best thing to do is to speak with the author. 

That said: you seem to be on Windows but not using Windows Terminal; is that correct? If so can you try this again with Windows Terminal.

**Reply by belonesox** [reply]
> Do note that that sample code has nothing to do with the Textual project. In the first instance, if you're having trouble with someone else's code, the best thing to do is to speak with the author.

Yes, first I [ask about it](https://github.com/davidfokkema/playground-textual/issues/1#issuecomment-1858991820) the author of the example, but looks that the example is trivial, and the problem in RichLog or something how to use it (but I failed to find another somehow working example of RichLog).

**Comment by belonesox** [comment]
Same story in standard Windows Terminal

`python -m rich.progress` looks OK
![image](https://github.com/Textualize/textual/assets/1609739/d65378d9-b188-411e-8bfb-d235612361a6)

`python console_richlog.py` looks NOK
![image](https://github.com/Textualize/textual/assets/1609739/57d22c62-1b82-4f9b-909b-ac43b386fa37)

**Reply by davep** [reply]
Are you certain that that's Windows Terminal you're using there? I don't have a Windows environment easily to hand to fully check, but that window in the image doesn't look like [Windows Terminal](https://apps.microsoft.com/detail/9N0DX20HK701?hl=en-gb&gl=US) to me.

**Comment by belonesox** [comment]
Same picture on new official Windows Terminal (not «cmd»)

`python -m rich.progress` looks OK

![image](https://github.com/Textualize/textual/assets/1609739/ac66c990-6ccb-4b22-a967-6716e7075c14)

`python console_richlog.py` looks NOK

![image](https://github.com/Textualize/textual/assets/1609739/2741bb93-6a56-4353-a79d-a20aeebb04b3)

**Reply by davep** [reply]
I think I'm still very confused about exactly what you're asking us here. As @TomJGooding pointed out above, `RichLog` works fine on Windows and you've confirmed that it's working just fine for you. You appear to be asking about how you can run some other person's code, but you're asking us about it rather than the author.

Moreover, now that I'm back from holiday and have access to a Python environment on a Windows machine, if I try and run the code you're asking about, it doesn't even appear to run as intended on Windows at all. When I run it I just get this:

![Screenshot 2024-01-04 at 09 11 56](https://github.com/Textualize/textual/assets/28237/27daff1e-3965-4fbf-8e36-761d068aa456)

That would make sense to me. The code being run is this:

```python
subprocess.Popen(
            "FORCE_COLOR=yes python -m rich.progress",
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            encoding="utf-8",
)
```

and that sort of command line isn't going to work when running something with the usual array of Windows shells. This would then suggest to me that you're running some sort of modified version of the code and you've not mentioned this, or there's something else about your environment that you've not mentioned.

But all of this again seems to come back to this: it would make far more sense for you to speak with the author of the code you're having an issue with; this doesn't seem to be a Textual issue as far as I can tell.

**Comment by davidfokkema** [comment, marked as answer]
Hi, original author of the `console_richlog.py` script here. Some context: this was a test to see if I could get coloured output from a subprocess into RichLog. This was not intended to be cross-platform at the moment, just trying out what pieces were in place and what was still missing. So, Rich (?) would detect it was not running in a terminal and would not output colour _at all_. Fixed with the environment variable, but that does not work on Windows like that. Then, RichLog does understand Rich markup, but _not_ ansi codes. So, had to use `Text.from_ansi()` which did work: on MacOS, I see coloured log entries in my RichLog! However, the Rich example moves the cursor back and you see updating progress indicators, but in RichLog every update is logged as a new log entry which looks _very_ different. I did not work out how to detect this and remove log entries. RichLog is really a _log_, and not a stand-in for a terminal window. So, that's the context.

Regarding this issue: RichLog _does_ support colour, both with the ansi roundtrip as shown by @TomJGooding, and also simply with `RichLog.write("[bold green]Hi there!")`. So... this issue can be closed?

**Reply by davep** [reply]
Thanks for the explanation David; appreciate it.

**Comment by belonesox** [comment]
Thanks all for answers, thanks @davidfokkema for `RichLog`, 
 
My intent is simple idea:
* I have a crossplatform textualize application that use some nice colors/progress features logging to terminal
* Now I want to make something like splitted areas (like `tmux`, but without `tmux`) adding some widgets to top area, and the log to bottom area.

```
|-------------------------------------------------------------|
| Widget1 |  Widget2 |  Widget3  |  Widget4  |  Widget5       | 
|-------------------------------------------------------------|
| .... the log from application                               |
| .... nice log ...                                           |
|  ..... with colors and all nice stuff  ..                   |
|-------------------------------------------------------------|
```

>RichLog is really a log, and not a stand-in for a terminal window.

Yes, `RichLog` is not ideal solution — if anyone know "how to split UI/log without tmux/screen", please let me know (may be I shoud ask other question).

But on first sight, looks, like `RichLog` — is the best solution for it, it almost "works for me" on linuxes!
Except: 
>However, the Rich example moves the cursor back and you see updating progress indicators

![image](https://github.com/Textualize/textual/assets/1609739/b5af699d-390d-4298-9b6f-14b6b2daf5c4)

But unfortunately, as you see, dont work on Windows.

> This was not intended to be cross-platform at the moment, just trying out what pieces were in place and what was still missing

Yes, it is the answer.  :(

So, may be an issue/RFE about "cross-platform/Windows support of RichLog" exists?  I would like to subscribe to it.

**Reply by TomJGooding** [reply]
The problem isn't with `RichLog`, so please stop suggesting it is.

> RichLog is really a log, and not a stand-in for a terminal window.

This is the answer, as you seem to recognise. The author of this `console_richlog.py` script has explained this was only an experiment - **it is not intended as a working example**.

If all you want is a progress bar, Textual has a widget for that. You don't need to involve logging Python subprocesses at all.

**Reply by davidfokkema** [reply]
Just to be clear, I did not write `RichLog`, I was merely trying it out and I'm running into the same issue you are, even on MacOS. If you run `python -m rich.progress` in a terminal, the cursor is moved back to the beginning of the lines so that the output seems to be updating in place. That does not work when you log the output to a log. A similar thing happens when you try to use a pager like `less` and use various flags to force the parsing of ANSI color codes. Try, on MacOS or Linux:
```
FORCE_COLOR=yes python -m rich.progress|less -R
```
That will also break, just like you see on Windows.

**Reply by davidfokkema** [reply]
Maybe head over to the Discord server, and ask your question in the #help-wanted channel? It is certainly possible to build the UI you want. What is difficult is to let your Textual application run _another_ application which is trying to do fancy terminal stuff and try to capture that output and include it in your application. What is _relatively easy_ is for your application to show some widgets and output some colourful text in a `RichLog`.

**Comment by davidfokkema** [comment]
Just a somewhat trivial example that does seem to implement your UI sketch:
<img width="842" alt="Screenshot 2024-01-04 at 22 41 45" src="https://github.com/Textualize/textual/assets/917137/f5b4cc2c-e97f-4ccc-9258-5b1cb68fc8ef">
This is the code:
```python
from textual import on
from textual.app import App, ComposeResult
from textual.containers import Horizontal
from textual.widgets import Button, RichLog


class LoggingApp(App[None]):
    CSS = """
        Horizontal {
            height: auto;
            align: center top;
        }

        Button, RichLog {
            margin: 1 2;
        }
    """

    def compose(self) -> ComposeResult:
        with Horizontal():
            yield Button("Green", id="green", variant="primary")
            yield Button("Red", id="red", variant="primary")
            yield Button("Blue", id="blue", variant="primary")
        yield RichLog(markup=True)

    @on(Button.Pressed)
    def log_text(self, event: Button.Pressed) -> None:
        color = event.button.id
        self.query_one("RichLog").write(f"Hi, I'm [bold {color}]{color}![/] :wink:")


if __name__ == "__main__":
    LoggingApp().run()
```
I tested that it works in the Terminal on Windows, ;-).

**Reply by belonesox** [reply]
>What is relatively easy is for your application to show some widgets and output some colourful text in a RichLog.

I know, yes.   But my application is Yet Another Build System, somehow like `rpmbuild`, where most of functionality on terminal-agnostic shell scripts, generated and called from Python Core, some of shell scripts show during compilation different types of progress-bars, coloring, etc, 
And I just want some Textual Widgets and Log from these application on one screen.

>Maybe head over to the Discord server, and ask your question in the #help-wanted channel?

I didn't want to strain the community by asking for immediate answers. I think the forums/discussions is a more appropriate tool for such questions.

>What is difficult is to let your Textual application run another application which is trying to do fancy terminal stuff and try to capture that >output and include it in your application

Now I got it, thank you for confirmation.   "a negative result is also useful"©.
May be someday, someone (may be me) implemented such use case in Textual.  `¯\_(ツ)_/¯`

**Reply by davidfokkema** [reply]
Then maybe this is what you're looking for: https://github.com/mitosch/textual-terminal? This wraps a full terminal emulator (using pyte) in a Textual widget. That code allows interacting with the terminal, which you maybe don't want if you just want to use it as a read-only log. In that case maybe you'll want to wrap pyte yourself, but textual-terminal shows you how to do that.
